### DISCUSSION TITLE
Terminal widget

### DISCUSSION THREAD

**Discussion by jgriffitts** [fltk-rs/fltk-rs#1498] [General]
For a month or so I've been working on the fltk-rs wrapper for the new `Fl_Terminal` widget (which replaces `Simple_Terminal`). At this point it works pretty well. **Is anyone else using Terminal in Rust?**

The Terminal API is large, and I haven't implemented interfaces for the whole thing. I believe that all public methods are supported, and some of the protected methods with priority for what I want for my own use cases. I am willing to take requests for other methods that are not yet there, and prioritize accordingly. Missing are
- `cursor_*` for cursor movement. These don't look useful to me and can be done with ANSI escape sequences. 
- `insert_*` and `delete_*` methods
- `selection_extend`, `u8c_cursor`, `w_to_col`, `walk_selection`, `vscroll_width`
- `Utf8Char` objects are immutable, supporting only getter methods, no setters. Essentially Utf8Char is only useful for looking into the RingBuffer arrays.

One other point: Terminal is now in the `fltk::group::experimental` module.**Is it time to take it out of `experimental`** and maybe out of `group`? I realize that in the C++ code Fl_Terminal is a subclass of Fl_Group, but this relationship won't be obvious at the Rust level.

**Comment by MoAlyousef** [comment]
Hello

Thank you for your contributions. I'm using it in fltk-term, which uses that in an editor I'm working on `red`.

I think FLTK is about to release version 1.4.0, which is currently in feature freeze (ongoing discussion [here](https://groups.google.com/g/fltkcoredev/c/SpR-dBJjn8Y), so hopefully soon). 1.4.0 will remove Fl_Simple_Terminal, which was supposed to be in 1.4.0 until work on Fl_Terminal started. I would prefer to wait until 1.4.0 is officially released before removing Terminal from the experimental module, since the api might still change and that might break users using fltk-rs major version 1.

My plan is that once 1.4.0 is released, I would create a last release for fltk-rs in major verison 1, then start working on version 2 which will have Terminal in its own module, and SimpleTerminal removed.

**Comment by jgriffitts** [comment]
I want to do a cleanup pass to track some changes in the Fl_Terminal API. Before I do this, I wanted to ask about the fltk-rs policy about protected methods. In my earlier work, 

The Fl_Terminal widget has a complex API with a lot of protected methods, and I built wrappers for a lot of those. A few of those are things I actually needed for my own use case. However I am afraid I may have gone overboard with exposing protected methods in the Rust API. Are there any general guidelines about this sort of thing?

**Reply by MoAlyousef** [reply]
Sure. Exposing protected methods is ok and sometimes necessary since Rust has no inheritance.
There are no guidelines, however if you feel a method is somehow complex or difficult to use correctly, you can mark it as `#[doc(hidden)]`.

**Reply by jgriffitts** [reply]
I believe I am done working on Fl_Terminal at this time. I don't expect to do anything else until/unless the fltk Terminal widget API changes, probably after the 1.4 release. However, if any issues come up with this widget, please feel free to assign them to me.

**Reply by MoAlyousef** [reply]
Will do. I appreciate your effort on the terminal widget. 
Thank you

**Comment by c61292558** [comment]
yes，I like this terminal
